string/string.cc

```diff
@@ -1,12 +1,154 @@
 #include "string.h"
 
-String::String()
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <cstring>
+
+const char* EMPTY = "";
+
+String::String() = default;
+
+String::~String() = default;
+
+String::String(String&& o) = default;
+
+String::String(const String& o): String(o.begin(), o.end()) {}
+
+String::String(const char* s): String(s, s + std::strlen(s)) {}
+
+template String::String(std::string::iterator first, std::string::iterator last);
+
```

Comment:
The definition of function template should be in the header file, so that all clients can 'see' it, and it would be automatically instantiated for all types required.
The the line above, instantiating for std::string::iterator wouldn't be needed.